logicrunner/handle_calls.go

```diff
@@ -0,0 +1,164 @@
+//
+// Copyright 2019 Insolar Technologies GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package logicrunner
+
+import (
+	"context"
+
+	"github.com/insolar/insolar/insolar"
+	"github.com/insolar/insolar/insolar/flow"
+	"github.com/insolar/insolar/insolar/flow/bus"
+	"github.com/insolar/insolar/insolar/message"
+	"github.com/insolar/insolar/insolar/reply"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/instrumentation/instracer"
+	"github.com/insolar/insolar/logicrunner/artifacts"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+type HandleCall struct {
+	dep *Dependencies
+
+	Message bus.Message
+}
+
+func (h *HandleCall) executeActual(
+	ctx context.Context,
+	parcel insolar.Parcel,
+	msg message.IBaseLogicMessage,
+	f flow.Flow) (insolar.Reply, error) {
+
+	lr := h.dep.lr
+	ref := msg.GetReference()
+	os := lr.UpsertObjectState(ref)
+
+	os.Lock()
+	if os.ExecutionState == nil {
+		os.ExecutionState = &ExecutionState{
+			Ref:   ref,
+			Queue: make([]ExecutionQueueElement, 0),
+		}
+	}
+	es := os.ExecutionState
+	os.Unlock()
+
+	// ExecutionState should be locked between CheckOurRole and
+	// appending ExecutionQueueElement to the queue to prevent a race condition.
+	// Otherwise it's possible that OnPulse will clean up the queue and set
+	// ExecutionState.Pending to NotPending. Execute will add an element to the
+	// queue afterwards. In this case cross-pulse execution will break.
+	es.Lock()
+
+	procCheckRole := CheckOurRole{
+		msg:  msg,
+		role: insolar.DynamicRoleVirtualExecutor,
+		Dep:  struct{ lr *LogicRunner }{lr: lr},
+	}
+
+	if err := f.Procedure(ctx, &procCheckRole); err != nil {
+		es.Unlock()
+		// TODO: check if error is ErrCancelled
+		return nil, errors.Wrap(err, "[ executeActual ] can't play role")
+	}
+
+	if lr.CheckExecutionLoop(ctx, es, parcel) {
+		es.Unlock()
+		return nil, os.WrapError(nil, "loop detected")
+	}
+	es.Unlock()
+
+	procRegisterRequest := NewRegisterRequest(parcel, h.dep)
+
+	if err := f.Procedure(ctx, procRegisterRequest); err != nil {
+		if err == flow.ErrCancelled {
+			request := procRegisterRequest.getResult()
```

Comment:
if pulse changed right before NewRegisterRequest we get error here just after start of f.Procedure and never get procRegisterRequest.getResult() because no one setResult()

Reply:
yes! I leave it as it was, without changes